### PR TITLE
Api

### PR DESCRIPTION
**Pull request by pardom-zz**
This PR includes the API interfaces written in Java. Other modules included are the core implementation `redux`, and a Kotlin extensions module `redux-kotlin`.

### CODE REVIEW

**Comment by beyondeye** [comment]
Hi Michael. It would be possible to not to nest the interface definition for subscription, enhancer and so on? What is your motivation (except for style preference)?

**Comment by pardom-zz** [comment]
The reason I made those internal is that they are concepts local to the `Store`, and not use in any other context. `Dispatcher`, `Reducer`, and `Middleware` are concepts beyond `Store`. Does that make sense, or should I give some examples to explain further?

**Comment by pardom-zz** [comment]
Your question brings up another point, however. `Middleware` should not be part of the API. The middleware API is an interface of the Middleware enhancer, which is an implementation of `Enhancer` and not a general concept of Redux.

**Comment by beyondeye** [comment]
Lets focus our discussion on Store.Subscriber and Store.Subscription, for example. I agree that  they are concepts related to the Store, but they are not internal. they can be referenced also from api consumer code. Also the class names are anbiguous because there are tons of classes/interfaces with the same name used in other contexts (for example rxjava). Also from an ide point of view, if you start typing in some place Subscriber you will not get automatically the right reference. Defining them at top level with the full name StoreSubscriber, StoreSubscription make it much easier when writing code to get the ide autocomplete get you the right reference

**Comment by pardom-zz** [comment]
Is this just a code style preference? I'm used to doing something like this, and my IDE has no problem handling it.

``` java
import rx.Subscriber;

class Foo {
    private Subscription subscription;

    void bar() {
        subscription = store.subscribe(Store.Subscriber {
            @Override 
            void onStateChanged() {
                // do stuff...
            }
        });
    }
}
```

Do you have any other reason for this other than code style? I'm not trying to block the conversation, I'm genuinely curious in exploring the benefits and drawbacks of either approach.

**Comment by beyondeye** [comment]
Enhancer and Creator are less problematic than Subscriber and Subscription (less ambigous), but similar arguments can apply also to them. So the choice is between easier coding (more help from the ide) and style preference (grouping all interfaces related to the store together as subclasses)

**Comment by pardom-zz** [comment]
I'm going to have a look at some other popular projects and see what they do, e.g. Square, Rx, etc.

**Comment by pardom-zz** [comment]
Also, while I'm open to considering the inner interfaces being extracted out, I think it would be a mistake to start prefixing classes with things like `Store`. Doing that obviates package names. For example, that pattern could be extrapolated to create classes like `ReduxStore` and `ReduxStoreSubscriber`. I think it makes more sense to use package qualification.

**Comment by beyondeye** [comment]
I agree, the problem is that Subscriber and Subscription are ambiguous. If you don't prefix something you will get from the ide a ton of different possible imports

**Comment by pardom-zz** [comment]
So I guess the question is do we prefer to qualify as `redux.Subscription` or `Store.Subscription`. Additionally, are there benefits and drawbacks to having the interfaces be internal or external.

I think moving them out of `Store` has the benefit of making them more visible to the consumer and promoting the concept of store subscription. We may want to do the same with `Enhancer`, in that case. However `Creator` is probably a concept that is only of concern to middleware/enhancer library creators.

**Comment by AngusMorton** [comment]
I prefer nesting the `Subscription` and `Subscriber` inside the Store. I think it provides a good way to qualify the `Subscription` as a store subscription as opposed to an `rx.Subscription`, for example.

I don't necessarily think visibility for the API consumer is a problem because autocomplete/documentation should make those types clear.

**Comment by beyondeye** [comment]
@AngusMorton Can you bring some example of a main java library that defines its main interfaces as nested interfaces? also take a look here: https://docs.oracle.com/javase/tutorial/java/javaOO/nested.html

File: api/src/main/java/redux/Store.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	 * @see <a href="http://redux.js.org/docs/api/Store.html#replaceReducer">http://redux.js.org/docs/api/Store.html#replaceReducer</a>
+	 */
+	void replaceReducer(Reducer<S> reducer);
+
+	/**
+	 * An interface that creates a Redux store.
+	 *
+	 * @see <a href="http://redux.js.org/docs/Glossary.html#store-creator">http://redux.js.org/docs/Glossary.html#store-creator</a>
+	 */
+	interface Creator<S> {
```
Comment (glung) [inline api/src/main/java/redux/Store.java]: The creator cannot cary the type, the reducer carries it. 
Otherwise, it is not possible to create an instrumented store for instance (because the caller sees type `S`When underneath it is actually `LiftedState<S>`)

My suggestion : 

``` java
interface Creator {

        /**
         *
         */
        <S> Store<S> create(Reducer<S> reducer, S initialState);
    }
```

**Review by glung — APPROVED** [review]
Great job 👍 

I requested only one change but I am giving my approval already to unlock you (because I'am going for a little holiday until Tuesday). 

About all the other comments, I don't really mind. This is an incremental work anyways, none of this is final. I trust you all to make the best informed decision at this time. 

My only request is that it is implementable by all stores out there with low frictions.

File: api/src/main/java/redux/Store.java (hunk 1 above)
Comment (beyondeye) [reply api/src/main/java/redux/Store.java]: 👍

**Comment by AngusMorton** [comment]
A `Subscriber` is effectively just a callback, like `OnClickListener` in Android or `OnImageLoaded` in Picasso. A `Subscription` is a bit different, but you could apply the same pattern to it for consistency.

I agree with Guillaume though in that I don't mind either way.

File: api/src/main/java/redux/Store.java (hunk 1 above)
Comment (pardom-zz) [reply api/src/main/java/redux/Store.java]: Hmm okay. I'll take a look at how this fits into my implementation so that I can understand it. I'll also take a look at your project and the JS source. I gave it a quick go and ran into typing issues, so I'll have to dig deeper to see how to make the typing work.

**Comment by pardom-zz** [comment]
I'm thinking about the implementation vs API classes at the moment. We'll probably want to create stores like `Store.createStore(reducer, state, middleware)`, but the `Store` class is already taken by the API. Should we package the API as `redux.internal.*` or something? Are there any other ways to solve this?

**Comment by pardom-zz** [comment]
Looking at the JS implementation again, the [`createStore`](http://redux.js.org/docs/api/createStore.html) function allows `initialState` to be optional. Thoughts on this? On the topic of nullability, we should annotate with `@NotNull` and `@Nullable` before merging this.

File: api/src/main/java/redux/Store.java (hunk 1 above)
Comment (pardom-zz) [reply api/src/main/java/redux/Store.java]: So after some investigation, I don't think moving the type to the function helps with lifting. You can do the same with the type on the interface. However, you _lose_ type safety for just about everything in the enhancer, and is the reason you need to cast in DevTools. The proposed change also messes with the typing in middleware. I'd be happy to create an example of lifting with the typed interface, if that helps.

Comment (pardom-zz) [reply api/src/main/java/redux/Store.java]: This is the gist of it though:

``` kotlin
class DevTools<S : Any> {

    data class State<S : Any>(val state: S)

    class LiftedStore<S> : Store<State<S>> {

        fun unlift(): State<S>

    }

    // etc, etc

}
```

Comment (glung) [reply api/src/main/java/redux/Store.java]: I am curious why do you think it is better to add the type to the creator ? 

To my understanding, a store creator knows how to create a store given a reducer and a state. I don't think it needs to carry a type.

I also don't think it is worth it because : 
- it is not the vanilla API (event if it is implicit) 
- we lose type safety (otherwise it [works fine](https://github.com/glung/jvm-redux-devtools-instrument/blob/master/lib/src/main/kotlin/redux/devtools/DevTools.kt#L71))
- It would work for your store and all stores out there (cf adapters here https://github.com/glung/jvm-redux)

**Comment by glung** [comment]
- I think we should try to avoid nullables until we prove it is needed. 
- I believe the API does not belong to an `internal`package because it is public by design. `redux.api`works wth me. 

Before merging it what do you think about proving it is implementable by other stores with low frictions ? I am happy to do it next week.

File: api/src/main/java/redux/Store.java (hunk 1 above)
Comment (beyondeye) [reply api/src/main/java/redux/Store.java]: Just a few comments from my personal experience. In reduks I use a lot Store Creators and I have them defined like pardom (`interface Creator<S>`). With hindsight it was kind of an  error, I had to add a method 

``` kotlin
 fun <S_> ofType(): StoreCreator<S_>
```

for making them work as I needed in my code. This is the reason I supported @glung request. It is true though, that not having the type info in interface itself create a lot of type safety problem when combining middlewares/store enhancers.
So perhaps, we can adopt the compromise I have in my code:

``` kotlin
interface StoreCreator<S> {
    /**
     * create a new store associated to this specific factory type
     */
    fun create(reducer: Reducer<S>, initialState: S):Store<S>
    /**
     * return new factory with same parameter but for new state type S_
     */
    fun <S_> ofType(): StoreCreator<S_>

}
```

**Comment by beyondeye** [comment]
>  We'll probably want to create stores like Store.createStore(reducer, state, middleware)

Not sure about this. In Reduks I have several types of Store: Simple, RxJava based, Promise based. Personally I don't feel the need to use the interface name also for the implementation, since there are several different implementations

File: redux-kotlin/build.gradle [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+apply plugin: 'kotlin'
+
+repositories {
+    mavenCentral()
+}
+
+dependencies {
+    compile project(':redux')
+
+    testCompile group: 'junit', name: 'junit', version: '4.11'
```
Comment (martin-g) [inline redux-kotlin/build.gradle]: 4.12

Comment (pardom-zz) [reply redux-kotlin/build.gradle]: 👍

File: api/src/main/java/redux/Store.java (hunk 1 above)
Comment (pardom-zz) [reply api/src/main/java/redux/Store.java]: Yeah I agree @beyondeye, moving the type to the `create` function makes the typing pointless. The store, the reducer, and the creator should all be typed by the state, because creator should always return a store using the same state type as the reducer.

My use of enhancers is pretty limited, so an example of an enhancer which needs to be differently typed would help me understand. I think my example above should work for any cases that I can think of so long as it can unlift to the original type.

Any examples I can take a run at?

Comment (beyondeye) [inline api/src/main/java/redux/Store.java]: @pardom about an example of my usage of ofType(). In my implementation I have the (non standard) concept of modules: i.e. combining multiple Stores into a single one. Basically think of the case in android where you have implemented the redux logic to drive the UI for two fragments (separate code for each fragment). With the concept of modules you can seamlessly integrate the redux code of the two fragments or even multiple instances of each fragment.
In order to do all this I need the Store Creator to allow store creation for multiple state classes
See 
- https://github.com/beyondeye/Reduks/blob/master/reduks/src/main/kotlin/com/beyondeye/reduks/modules/ReduksModule.kt
- https://github.com/beyondeye/Reduks/blob/master/reduks/src/test/kotlin/com/beyondeye/reduks/modules/MultiReduksTest.kt

Basically I try to overcome the limitations of combineReducers vs the original ReduxJs version.

Comment (beyondeye) [reply api/src/main/java/redux/Store.java]: Never mind 😄 .I've found a way to refactor code so that I would not need ofType() in Creator interface

Comment (pardom-zz) [reply api/src/main/java/redux/Store.java]: Okay, so I want to make sure everyone is on the same page about this before merging. 
- Keeping the type on `Creator` is the only way to enforce type safety.
- Lifting/delegating the base store requires no typing on the function
- No additional typing is required to accomplish lifting/delegating

Since @glung brought this up originally, I want to make sure your concerns are addressed too. I can update your `devtools-instrument` to work with this interface if you'd like.

Comment (glung) [reply api/src/main/java/redux/Store.java]: @pardom as discussed, here is an example to illustrate my point : https://github.com/jvm-redux/jvm-redux-api/blob/api-store-creator-experiment/api/src/main/java/example/SomeEnhancer.java#L18

My thinking is : 
- the reducer reduces a type S
- the creator holds the type reduced by the reducer. 

Hence, a creator can create a store of the type the reducer reduces.  And I believe it is what we need to make it possible to create a dev store.

``` java
Creator {
<S> Store<S> create(Reducer<S> reducer); 
|
```

A bit hairy, I know. I hope the example will bring some clarity

File: api/src/main/java/redux/api/Store.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	interface Creator<S> {
+
+		/**
+		 * Creates a store.
+		 *
+		 * @param reducer Reducer for the store state
+		 * @param initialState Initial state for the store
+		 * @return The store
+		 */
+		Store<S> create(Reducer<S> reducer, S initialState);
```
Comment (glung) [inline api/src/main/java/redux/api/Store.java]: I commented on the close conversation 
https://github.com/jvm-redux/jvm-redux-api/pull/2#discussion_r84126134

Maybe let's continue here.

Comment (pardom-zz) [reply api/src/main/java/redux/api/Store.java]: @beyondeye, @glung and I just chatted about this gist: https://gist.github.com/pardom/21124eca191b092ed1a886e2452f00a6

It demonstrates the type enforcement I was referring to, while still allowing for an enhanced State/Store.

Comment (glung) [reply api/src/main/java/redux/api/Store.java]: I see. 

I guess, it boils down to programming style : wrapping the store is more oo, and lifting the the reducer is more functional 

In the end, because you demonstrated there is not limitations with this approach, from my side we can go with it.

I am curious though about @beyondeye thoughts ?

**Comment by pardom-zz** [comment]
@beyondeye @glung LGTY?

**Comment by beyondeye** [comment]
[LGTY?](http://www.meaningfor.com/what-does-stand-for/lgty)

**Comment by pardom-zz** [comment]
LGTY? = Looks good to you?

**Comment by beyondeye** [comment]
Just joking. For me looks OK. Need to find some time and try integrate with reduks, but I assume we can fix things on the go if something goes terribly wrong

**Comment by glung** [comment]
\o/
